### DISCUSSION TITLE
Convergence issue while solving phase-field fracture

### DISCUSSION THREAD

**Discussion by avtarsinghh1991** [idaholab/moose#20016] [Q&A Modules: General] [answered]
Hello MOOSE Experts

I am trying to solve rate-independent phase-field fracture using the following example

`crack2d_iso_wo_time.i`

for CT specimen geometry, the material properties are taken from Miehe's paper.

However, I am getting the following error as soon as the damage evolves to value 1.

```
  Linear solve did not converge due to DIVERGED_ITS iterations 10
  Linear solve did not converge due to DIVERGED_ITS iterations 10
  Linear solve did not converge due to DIVERGED_ITS iterations 10
  Linear solve did not converge due to DIVERGED_ITS iterations 10
  Linear solve did not converge due to DIVERGED_ITS iterations 10
  Linear solve did not converge due to DIVERGED_ITS iterations 10
  Linear solve did not converge due to DIVERGED_ITS iterations 10
  Linear solve did not converge due to DIVERGED_ITS iterations 10
  Linear solve did not converge due to DIVERGED_ITS iterations 10
  Linear solve did not converge due to DIVERGED_ITS iterations 10
Nonlinear solve did not converge due to DIVERGED_MAX_IT iterations 10
 Solve Did NOT Converge!
```
![Untitled](https://user-images.githubusercontent.com/94848840/149025090-69d3cedd-1402-4025-9587-f31dc142f187.png)

**Input file I used is as follows;**

```
#This input does not add time derivative kernel for phase field equation
[Mesh]
  file = 'Miehe.inp'
[]

[GlobalParams]
  displacements = 'disp_x disp_y'
[]

[Modules]
  [./TensorMechanics]
    [./Master]
      [./mech]
        add_variables = true
        strain = SMALL
        additional_generate_output = 'stress_yy'
        save_in = 'resid_x resid_y'
      [../]
    [../]
  [../]
[]

[Variables]
  [./c]
    order = FIRST
    family = LAGRANGE
  [../]
[]

[AuxVariables]
  [./resid_x]
  [../]
  [./resid_y]
  [../]
[]

[Kernels]
  [./solid_x]
    type = PhaseFieldFractureMechanicsOffDiag
    variable = disp_x
    component = 0
    c = c
  [../]
  [./solid_y]
    type = PhaseFieldFractureMechanicsOffDiag
    variable = disp_y
    component = 1
    c = c
  [../]
  [./ACBulk]
    type = AllenCahn
    variable = c
    f_name = F
  [../]

  [./ACInterface]
    type = ACInterface
    variable = c
    kappa_name = kappa_op
  [../]
[]

[BCs]
  [./ydisp]
    type = FunctionDirichletBC
    variable = disp_y
    boundary = NS4
    function = 't'
  [../]
    [./xdisp]
    type = DirichletBC
    variable = disp_x
    boundary = NS4
    value = 0
  [../]
  [./yfix]
    type = DirichletBC
    variable = disp_y
    boundary = NS3
    value = 0
  [../]
  [./xfix]
    type = DirichletBC
    variable = disp_x
    boundary = NS3
    value = 0
  [../]
[]

[Materials]
  [./pfbulkmat]
    type = GenericConstantMaterial
    prop_names = 'gc_prop l visco'
    prop_values = '2.7 0.015 3.0e-6'
  [../]
  [./define_mobility]
    type = ParsedMaterial
    material_property_names = 'gc_prop visco'
    f_name = L
    function = '1.0/(gc_prop * visco)'
  [../]
  [./define_kappa]
    type = ParsedMaterial
    material_property_names = 'gc_prop l'
    f_name = kappa_op
    function = 'gc_prop * l'
  [../]
  [./elasticity_tensor]
    type = ComputeElasticityTensor
    C_ijkl = '121.0e3 80.0e3'
    fill_method = symmetric_isotropic
  [../]
  [./elastic]
    type = ComputeLinearElasticPFFractureStress
    c = c
    E_name = 'elastic_energy'
    D_name = 'degradation'
    F_name = 'local_fracture_energy'
    decomposition_type = strain_spectral
  [../]
  [./degradation]
    type = DerivativeParsedMaterial
    f_name = degradation
    args = 'c'
    function = '(1.0-c)^2*(1.0 - eta) + eta'
    constant_names       = 'eta'
    constant_expressions = '0.0'
    derivative_order = 2
  [../]
  [./local_fracture_energy]
    type = DerivativeParsedMaterial
    f_name = local_fracture_energy
    args = 'c'
    material_property_names = 'gc_prop l'
    function = 'c^2 * gc_prop / 2 / l'
    derivative_order = 2
  [../]
  [./fracture_driving_energy]
    type = DerivativeSumMaterial
    args = c
    sum_materials = 'elastic_energy local_fracture_energy'
    derivative_order = 2
    f_name = F
  [../]
[]

[Postprocessors]
  [./resid_x]
    type = NodalSum
    variable = resid_x
    boundary = NS3
  [../]
  [./resid_y]
    type = NodalSum
    variable = resid_y
    boundary = NS3
  [../]
[]

[Preconditioning]
  [./smp]
    type = SMP
    full = true
  [../]
[]

[Executioner]
  type = Transient

  automatic_scaling = true
  compute_scaling_once=false
  solve_type = 'PJFNK'
  petsc_options_iname = '-pc_type -pc_hypre_type -ksp_gmres_restart'
  petsc_options_value = 'hypre boomeramg 101'

  nl_rel_tol = 1e-3
  l_tol = 1e-3

  l_max_its = 10
  nl_max_its = 10

  dt = 1e-5
  dtmin = 1e-6
  num_steps = 3500
[]

[Outputs]
  exodus = true
  interval = 10
[]
```

For reference. I have also uploaded the distribution of the damage field at the time just before the system diverged. 

Can you please let me know the procedure to resolve this instability issue?

I shall be very thankful.

Best
Avtar

**Comment by GiudGiud** [comment]
Hello

Can you please give us the full output log?
l_max_its of 10 is not a lot, this is triggering `Linear solve did not converge due to DIVERGED_ITS iterations 10`, which could be taking down the non linear solver (since linear iterates are not very converged)

Guillaume

**Reply by avtarsinghh1991** [reply]
Please find the attached file. I increased the linear solver iterations to 50. Still, it's getting crashed at approx the same load step.

[output.zip](https://github.com/idaholab/moose/files/7851222/output.zip)

**Comment by ngrilli** [comment, marked as answer]
Dear @asingh-mit 

Try to start from the setup of the example with the variational inequality solver, which is well converging:
https://github.com/idaholab/moose/blob/next/modules/combined/test/tests/phase_field_fracture/crack2d_vi_solver.i

Other convergence issues can be due to the load applied to quickly or the values of "l" and "visco"
Make sure "l" is about 4 times your mesh size, at least

The only thing I can see in your input file is:
  nl_rel_tol = 1e-3
  l_tol = 1e-3
These are too high, just leave them as default.

Also increase a bit number of linear and non-linear iterations
as suggested by @GiudGiud

You can see other examples of similar converging simulations here:
https://github.com/ngrilli/c_pfor_am/tree/main/test/tests/CPDamage

Best Regards,
Nicolò

**Reply by GiudGiud** [reply]
Please follow @ngrilli 's advice first. And if it unfortunately doesnt work, please reduce your mesh to 1000 dofs or less and run with `-pc_type svd -pc_svd_monitor ` on the command line. This will let us know if the system becomes singular right when you are getting convergence issues.

**Reply by jiangwen84** [reply]
Add one more thing: If your linear solver fails, try to use a direct solver. 
 ```  petsc_options_iname = '-pc_type'``` 
 ``` petsc_options_value = 'lu'```

**Reply by avtarsinghh1991** [reply]
Thank you very much for the suggestions. Let you know the results after the simulation.

Hey @ngrilli 
As you suggested about the load applied, I completely agree.
So, I am wondering if there is any method to decrease it after certain load steps.

Let's say,
I want to apply displacement of 1.0e-5 for the first 500 and then 1.0e-6 for the rest of the simulation.

Best
Avtar

**Reply by jiangwen84** [reply]
Take a look at `TimeSteppers`  https://mooseframework.inl.gov/syntax/Executioner/TimeStepper/index.html

**Comment by avtarsinghh1991** [comment]
Hello MOOSE Experts

I followed the suggestions given by @ngrilli and @jiangwen84. But the system is still diverging as soon as the damage value reaches 1. Next, as suggested by @GiudGiud, I will decrease the problem size and provide the output.

I just have one doubt. I need to incorporate anisotropy in the phase-field fracture and also the rate-independent. Is it possible to do that or do I need to write the kernel for the same?

Best
Avtar

**Reply by ngrilli** [reply]
Dear @asingh-mit 

Please attach the new input file, otherwise it's difficult to say.
Are you setting any initial condition for the damage field? Or does it start from 0 everywhere?
My experience is that you need to initialise the phase field with a function exp(-y/l)
so mimicking the steady state solution, otherwise you get convergence problems at the very beginning.
If your initial phase field is zero, there should not be problems.

Yes, anisotropic phase field fracture, to see anisotropic crack propagation, is already available in MOOSE:
https://github.com/idaholab/moose/blob/next/modules/combined/test/tests/phase_field_fracture/crack2d_aniso_cleavage_plane.i
no need to write any kernel.

There is one further 3D example here:
https://github.com/ngrilli/c_pfor_am/blob/main/test/tests/CPDamage/AnisoCleavage.i
I made this and I am sure this is converging well and show crack propagation.

The fracture surface is a constant vector defined by "cleavage_plane_normal"

Best Regards,
Nicolò

**Reply by avtarsinghh1991** [reply]
Hello @ngrilli @jiangwen84 @GiudGiud 

Thank you very much for your help.  I successfully validated the uniaxial tension test problem given in Prof. Miehe's Paper titled "**Thermodynamically consistent phase-field models of fracture: Variational principles and multi-field FE implementations**". 

I have tried many solvers and preconditioners settings. I can only be able to solve with the following solver settings:

```
  solve_type = NEWTON
  petsc_options_iname = '-pc_type -pc_factor_mat_solver_package -snes_type'
  petsc_options_value = 'lu       superlu_dist                  vinewtonrsls'
  automatic_scaling = true
```

Other solvers like PJFNK diverge as soon as the damage value reaches 1 or load started to drop. I have attached the two figures. First with PJFNK solver and second with Newton solver.

Best
Avtar

![as](https://user-images.githubusercontent.com/94848840/149781341-0d1caebe-1aeb-47b8-a91d-df97b6f5d237.jpg)
![Untitled](https://user-images.githubusercontent.com/94848840/149781358-e93c85fd-b4a5-4553-8898-cc65a84303ff.jpg)

**Reply by jiangwen84** [reply]
`vinewtonrsls` should work with PJFNK too, but not sure why it fails to converge for this case. 

Yes, we typically use direct solver for mechanics/fracture problems, at least for 2D and small 3D cases.
